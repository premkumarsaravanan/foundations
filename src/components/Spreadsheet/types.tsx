import * as React from 'react'
import ReactDataSheet from 'react-datasheet'

export class MyReactDataSheet extends ReactDataSheet<Cell, string> {}

/** Cell contain predefined value
 * https://github.com/nadbm/react-datasheet/blob/master/types/react-datasheet.d.ts
 * plus some properties below
 */
export interface Cell extends ReactDataSheet.Cell<Cell, string> {
  /** The value of the cell, always a string */
  value: string
  /** The validate function, receive Cell as param, must return boolean */
  validate?: (cell: Cell) => boolean
  /** Additional className for styling cell */
  className?: string
  style?: React.CSSProperties
  CustomComponent?: React.FC<{
    data: Cell[][]
    cellRenderProps: ReactDataSheet.CellRendererProps<Cell>
    setData: React.Dispatch<Cell[][]>
    setSelected: React.Dispatch<SelectedMatrix>
  }>
}

export interface CellWithValidate extends Cell {
  validate: (cell: Cell) => boolean
}

export interface DoubleClickPayLoad {
  row: number
  col: number
  maxRowIndex: number
  maxColIndex: number
  isReadOnly?: boolean
}

export interface SpreadsheetProps {
  data: Cell[][]
  description?: React.ReactNode
  hasUploadButton?: boolean
  hasDownloadButton?: boolean
  hasAddButton?: boolean
  /** This will run before data is set to table, must return data with validate function */
  validateUpload?: (data: Cell[][]) => CellWithValidate[][]
}

export interface SelectedMatrix {
  start: ReactDataSheet.Location
  end: ReactDataSheet.Location
}
