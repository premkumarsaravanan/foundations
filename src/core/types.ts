export interface LoginSession {
  accessToken: string
  accessTokenExpiry: number
  idToken: string
  idTokenExpiry: number
  refreshToken: string
}

export interface LoginParams {
  userName: string
  password: string
}

export interface ChangePasswordParams {
  newPassword: string
  userName: string
  password: string
}

export interface RefreshParams {
  userName: string
  refreshToken: string
}

export interface ConfirmPasswordParams {
  userName: string
  newPassword: string
  verificationCode: string
}

export interface ConfirmRegistrationParams {
  userName: string
  verificationCode: string
}

export interface ResetPasswordParams {
  userName: string
}
