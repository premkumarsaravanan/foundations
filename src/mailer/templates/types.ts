export interface ForgottenPasswordTemplateParams {
  verificationCode: string
  userName: string
}

export interface ConfirmPasswordTemplateParams {
  userName: string
}
