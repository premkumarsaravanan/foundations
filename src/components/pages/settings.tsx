import * as React from 'react'

import { FlexContainerResponsive, Content, H3, H4, FlexContainerBasic } from '@reapit/elements'

const SettingsPage: React.SFC = () => {
  return (
    <FlexContainerBasic flexColumn hasPadding>
      <Content>
        <FlexContainerResponsive flexColumn hasBackground hasPadding>
          <H3>Settings</H3>
          <H4>Payment Info</H4>
        </FlexContainerResponsive>
      </Content>
    </FlexContainerBasic>
  )
}

export default SettingsPage
