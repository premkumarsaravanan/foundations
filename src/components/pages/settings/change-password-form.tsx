import React from 'react'
import { Form, withFormik, FormikProps, FormikBag, FormikErrors } from 'formik'
import { compose } from 'redux'
import {
  FlexContainerResponsive,
  GridItem,
  FormSection,
  FormHeading,
  FormSubHeading,
  Grid,
  Input,
  Button
} from '@reapit/elements'
import { withRouter, RouterProps } from 'react-router'

export type ChangePasswordFormProps = FormikProps<{}>

export const ChangePasswordForm: React.FC<ChangePasswordFormProps> = ({ isSubmitting, isValidating, isValid }) => {
  return (
    <FormSection>
      <Form>
        <FormHeading>Change password</FormHeading>
        <FormSubHeading>
          Please complete the following fields to change your password. You will be automatically logged out for the
          changed to be applied
        </FormSubHeading>
        <Grid>
          <GridItem>
            <Input
              dataTest="current-password"
              type="password"
              labelText="Current Password"
              id="currentPassword"
              name="currentPassword"
            />
          </GridItem>
        </Grid>
        <Grid>
          <GridItem>
            <Input dataTest="password" type="password" labelText="Password" id="password" name="password" />
          </GridItem>
          <GridItem>
            <Input
              dataTest="confirmPassword"
              type="password"
              labelText="Confirm Password"
              id="confirmPassword"
              name="confirmPassword"
            />
          </GridItem>
        </Grid>
        <FlexContainerResponsive>
          <Button disabled={!isValid} loading={isSubmitting || isValidating} variant="primary" type="submit">
            Change Password
          </Button>
        </FlexContainerResponsive>
      </Form>
    </FormSection>
  )
}

export type ChangePasswordValues = {
  currentPassword: string
  password: string
  confirmPassword: string
}

export const validateChangePasswordForm = (values: ChangePasswordValues) => {
  const errors: FormikErrors<ChangePasswordValues> = {}
  if (values.password !== values.confirmPassword) {
    errors.confirmPassword = 'Passwords do not match'
  }
  return errors
}

export const mapPropsChangePassword = () => ({
  currentPassword: '',
  password: '',
  confirmPassword: ''
})

export const changePassword = (values: ChangePasswordValues) => {
  console.log(values)
  // TODO: will replace mock here for real API
  return new Promise((resolve, reject) => {
    setTimeout(() => {
      resolve(true)
    }, 500)
    setTimeout(() => {
      reject(new Error('some errors'))
    }, 500)
  })
}

export type EnhanceChangePasswordFormProps = {
  logout: () => void
  errorNotification: () => void
}

export const handleSubmitChangePassword = async (
  values: ChangePasswordValues,
  { setSubmitting, props }: FormikBag<EnhanceChangePasswordFormProps & RouterProps, ChangePasswordValues>
) => {
  try {
    const response = await changePassword(values)
    if (response) {
      setSubmitting(false)
      props.logout()
      const SUCCESS_ALERT_LOGIN_PAGE = '/developer/login?isChangePasswordSuccess=1'
      props.history.replace(SUCCESS_ALERT_LOGIN_PAGE)
    }
  } catch (error) {
    console.error(error)
    props.errorNotification()
  }
}

export const withChangePasswordForm = withFormik({
  displayName: 'WithChangePasswordForm',
  validate: validateChangePasswordForm,
  mapPropsToValues: mapPropsChangePassword,
  handleSubmit: handleSubmitChangePassword
})

const EnhanceChangePasswordForm = compose<React.FC<EnhanceChangePasswordFormProps>>(
  withRouter,
  withChangePasswordForm
)(ChangePasswordForm)
EnhanceChangePasswordForm.displayName = 'EnhanceChangePasswordForm'

export default EnhanceChangePasswordForm
