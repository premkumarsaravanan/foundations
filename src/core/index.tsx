import '../styles/index.scss'
import * as React from 'react'
import { render } from 'react-dom'
import Router from './router'
import { Provider } from 'react-redux'
import store from './store'
import Toast from '../components/ui/toast'
import { PortalProvider } from '@reapit/elements'

const rootElement = document.querySelector('#root') as Element

const App = () => (
  <Provider store={store.reduxStore}>
    <PortalProvider>
      <Router />
    </PortalProvider>
    <Toast />
  </Provider>
)

if (rootElement) {
  render(<App />, rootElement)
}

export default App
