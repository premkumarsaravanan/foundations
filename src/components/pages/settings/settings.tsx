import * as React from 'react'
import {
  FlexContainerResponsive,
  Content,
  H3,
  FlexContainerBasic,
  GridItem,
  Grid,
  Loader,
  fetcher
} from '@reapit/elements'
import { URLS, MARKETPLACE_HEADERS } from '@/constants/api'
import EnhanceContactInformation, { ContactInformationValues } from './contact-information-form'
import EnhanceChangePasswordForm, { ChangePasswordValues } from './change-password-form'
import { compose, Dispatch } from 'redux'
import { connect } from 'react-redux'
import { ReduxState } from '@/types/core'
import { errorThrownServer } from '@/actions/error'
import errorMessages from '@/constants/error-messages'
import { authClear } from '@/actions/auth'
import { DeveloperModel } from '@reapit/foundations-ts-definitions'
import { updateDeveloperData, changePassword } from '@/actions/settings'

export type SettingsPageProps = StateProps & DispatchProps

export const SettingsPage: React.FC<SettingsPageProps> = ({
  developerInfo,
  email,
  loading,
  updateDeveloperInformation,
  changePassword
}) => {
  if (loading) {
    return <Loader />
  }
  return (
    <FlexContainerBasic flexColumn hasPadding>
      <Content>
        <FlexContainerResponsive flexColumn hasBackground hasPadding>
          <H3>Settings</H3>
          <Grid>
            <GridItem>
              <EnhanceContactInformation
                developerInformation={developerInfo}
                updateDeveloperInformation={updateDeveloperInformation}
              />
            </GridItem>
            <GridItem>
              <EnhanceChangePasswordForm email={email} changePassword={changePassword} />
            </GridItem>
          </Grid>
        </FlexContainerResponsive>
      </Content>
    </FlexContainerBasic>
  )
}

export type StateProps = {
  developerInfo: DeveloperModel | null
  email: string
  loading: boolean
}

export const mapStateToProps = (state: ReduxState): StateProps => {
  return {
    developerInfo: state.settings.developerInfomation || {},
    email: state.auth?.loginSession?.loginIdentity?.email || '',
    loading: state.settings?.loading
  }
}

export type DispatchProps = {
  updateDeveloperInformation: (values: ContactInformationValues) => void
  changePassword: (values: ChangePasswordValues) => void
}

export const mapDispatchToProps = (dispatch: Dispatch): DispatchProps => {
  return {
    updateDeveloperInformation: (values: ContactInformationValues) => dispatch(updateDeveloperData(values)),
    changePassword: (values: ChangePasswordValues) => dispatch(changePassword(values))
  }
}

export const withRedux = connect(mapStateToProps, mapDispatchToProps)

export const EnhanceSettingPage = compose<React.FC<SettingsPageProps>>(withRedux)(SettingsPage)
EnhanceSettingPage.displayName = 'EnhanceSettingPage'

export default EnhanceSettingPage as React.LazyExoticComponent<any>
