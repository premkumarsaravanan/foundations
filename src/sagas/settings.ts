import { put, fork, all, call, takeLatest, select } from '@redux-saga/core/effects'
import { fetcher, removeSessionCookie } from '@reapit/elements'
import { Action } from '@/types/core'
import ActionTypes from '@/constants/action-types'
import errorMessages from '@/constants/error-messages'
import { MARKETPLACE_HEADERS, REAPIT_API_BASE_URL, URLS, COGNITO_API_BASE_URL } from '@/constants/api'
import { settingShowLoading, requestDeveloperDataSuccess, ChangePasswordParams } from '@/actions/settings'
import { errorThrownServer } from '@/actions/error'
import { DeveloperModel } from '@/types/marketplace-api-schema'
import { selectDeveloperId, selectDeveloperEmail } from '@/selector/developer'
import { authLogoutSuccess } from '@/actions/auth'
import Routes from '@/constants/routes'
import { history } from '../core/router'

export const fetchDeveloperInfo = async (developerId: string | null | undefined) => {
  const response = await fetcher({
    url: `${URLS.developers}/${developerId}`,
    api: REAPIT_API_BASE_URL,
    method: 'GET',
    headers: MARKETPLACE_HEADERS
  })
  return response
}

export const developerInformationFetch = function*() {
  yield put(settingShowLoading(true))
  try {
    const developerId = yield select(selectDeveloperId)
    if (!developerId) {
      return
    }
    const response = yield call(fetchDeveloperInfo, developerId)
    if (response) {
      yield put(requestDeveloperDataSuccess(response))
    }
  } catch (error) {
    console.error(error)
    yield put(
      errorThrownServer({
        type: 'SERVER',
        message: errorMessages.DEFAULT_SERVER_ERROR
      })
    )
  } finally {
    yield put(settingShowLoading(false))
  }
}

export type UpdateDeveloperInfoParams = {
  developerId: string
  values: DeveloperModel
}

export const updateDeveloperInfo = async ({ developerId, values }: UpdateDeveloperInfoParams) => {
  const response = await fetcher({
    url: `${URLS.developers}/${developerId}`,
    api: REAPIT_API_BASE_URL,
    method: 'PUT',
    headers: MARKETPLACE_HEADERS,
    body: values
  })
  return response
}

export const developerInfomationChange = function*({ data }: Action<DeveloperModel>) {
  yield put(settingShowLoading(true))
  try {
    const developerId = yield select(selectDeveloperId)
    if (!developerId) {
      return
    }
    const response = yield call(updateDeveloperInfo, { developerId, values: data })
    if (response) {
      const newResponse = yield call(fetchDeveloperInfo, developerId)
      yield put(requestDeveloperDataSuccess(newResponse))
    }
  } catch (error) {
    console.error(error)
    yield put(
      errorThrownServer({
        type: 'SERVER',
        message: errorMessages.DEFAULT_SERVER_ERROR
      })
    )
  } finally {
    yield put(settingShowLoading(false))
  }
}

export type CallChangePasswordParams = {
  values: ChangePasswordParams
  email: string
}

export const callChangePassword = async ({ values, email }: CallChangePasswordParams) => {
  const CHANGE_PASSWORD_URL = '/password/change'
  const response = await fetcher({
    url: CHANGE_PASSWORD_URL,
    api: COGNITO_API_BASE_URL,
    method: 'POST',
    headers: MARKETPLACE_HEADERS,
    body: {
      password: values.currentPassword,
      newPassword: values.password,
      userName: email
    }
  })
  return response
}

export const developerPasswordChange = function*({ data }: Action<ChangePasswordParams>) {
  yield put(settingShowLoading(true))
  try {
    const email = yield select(selectDeveloperEmail)
    const response = yield call(callChangePassword, { values: data, email })
    const isCallAPISuccess = response.message === 'SUCCESS'
    if (isCallAPISuccess) {
      const SUCCESS_ALERT_LOGIN_PAGE = `${Routes.DEVELOPER_LOGIN}?isChangePasswordSuccess=1`
      yield call(removeSessionCookie)
      yield history.replace(SUCCESS_ALERT_LOGIN_PAGE)
      yield put(authLogoutSuccess())
    }
  } catch (error) {
    yield put(
      errorThrownServer({
        type: 'SERVER',
        message: errorMessages.DEFAULT_SERVER_ERROR
      })
    )
  } finally {
    yield put(settingShowLoading(false))
  }
}

export const developerInformationFetchListen = function*() {
  yield takeLatest<Action<string>>(ActionTypes.SETTING_FETCH_DEVELOPER_INFO, developerInformationFetch)
}

export const developerInformationChangeListen = function*() {
  yield takeLatest<Action<DeveloperModel>>(ActionTypes.SETTING_UPDATE_DEVELOPER, developerInfomationChange)
}

export const developerPasswordChangeListen = function*() {
  yield takeLatest<Action<ChangePasswordParams>>(ActionTypes.CHANGE_PASSWORD, developerPasswordChange)
}

export const settingsSagas = function*() {
  yield all([
    fork(developerInformationFetchListen),
    fork(developerInformationChangeListen),
    fork(developerPasswordChangeListen)
  ])
}

export default settingsSagas
