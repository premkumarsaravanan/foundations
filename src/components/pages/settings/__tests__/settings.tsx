import React from 'react'
import { shallow } from 'enzyme'
import { SettingsPageProps, SettingsPage, mapStateToProps, handleUseEffect, mapDispatchToProps } from '../settings'
import { ReduxState } from '@/types/core'

const mockDeveloperInformation = {
  id: '7b2517b3-aad3-4ad8-b31c-988a4b3d112d',
  externalId: 'ecaf36e1-2b4c-46c2-90d8-1cb493e5dcbd',
  name: 'Reapit Ltd',
  company: 'Reapit Ltd',
  jobTitle: 'Head of Cloud',
  email: 'wmcvay@reapit.com',
  telephone: '01234 567890',
  created: '2019-07-31T11:34:30',
  modified: '2019-11-18T08:04:52'
}

jest.mock('@reapit/elements', () => ({
  fetcher: jest.fn().mockResolvedValue(mockDeveloperInformation)
}))

describe('SettingsPage', () => {
  it('should match snapshot', () => {
    const mockProps: SettingsPageProps = {
      developerId: '7b2517b3-aad3-4ad8-b31c-988a4b3d112d',
      logout: jest.fn(),
      errorNotification: jest.fn()
    }
    const wrapper = shallow(<SettingsPage {...mockProps} />)
    expect(wrapper).toMatchSnapshot()
  })
  it('should match snapshot', () => {
    const mockProps: SettingsPageProps = {
      developerId: null,
      logout: jest.fn(),
      errorNotification: jest.fn()
    }
    const wrapper = shallow(<SettingsPage {...mockProps} />)
    expect(wrapper).toMatchSnapshot()
  })
  describe('handleUseEffect', () => {
    it('should run correctly', done => {
      const mockProps = { developerId: '123', setDeveloperInformation: jest.fn(), setLoading: jest.fn() }
      const fn = handleUseEffect(mockProps)
      fn()
      setTimeout(() => {
        expect(mockProps.setDeveloperInformation).toBeCalled()
        expect(mockProps.setLoading).toBeCalled()
        done()
      }, 100)
    })
  })
  describe('mapStateToProps', () => {
    it('should run correctly', () => {
      const mockState = {
        auth: {
          loginSession: {
            loginIdentity: {
              developerId: '123'
            }
          }
        }
      } as ReduxState
      const output = {
        developerId: mockState.auth?.loginSession?.loginIdentity.developerId
      }
      const result = mapStateToProps(mockState)
      expect(result).toEqual(output)
    })

    it('should return null', () => {
      const mockState = {
        auth: {}
      } as ReduxState
      const output = {
        developerId: null
      }
      const result = mapStateToProps(mockState)
      expect(result).toEqual(output)
    })
  })
  describe('mapDispatchToProps', () => {
    it('should call dispatch when logout', () => {
      const mockDispatch = jest.fn()
      const { logout } = mapDispatchToProps(mockDispatch)
      logout()
      expect(mockDispatch).toBeCalled()
    })

    it('should call dispatch when errorNotification', () => {
      const mockDispatch = jest.fn()
      const { errorNotification } = mapDispatchToProps(mockDispatch)
      errorNotification()
      expect(mockDispatch).toBeCalled()
    })
  })
})
