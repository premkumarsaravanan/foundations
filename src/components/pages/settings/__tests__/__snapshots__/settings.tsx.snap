// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsPage should match snapshot 1`] = `<Loader />`;

exports[`SettingsPage should match snapshot 2`] = `
<FlexContainerBasic
  flexColumn={true}
  hasPadding={true}
>
  <Content>
    <FlexContainerResponsive
      flexColumn={true}
      hasBackground={true}
      hasPadding={true}
    >
      <H3>
        Settings
      </H3>
      <Grid>
        <GridItem>
          <EnhanceContactInformation
            developerInformation={
              Object {
                "company": "Dwarves Foundation",
                "created": "2019-11-21T08:46:41",
                "email": "tanphamhaiduong@gmail.com",
                "externalId": "6447271a-0268-4546-b628-878308d091eb",
                "id": "7a96e6b2-3778-4118-9c9b-6450851e5608",
                "isInactive": false,
                "jobTitle": "Software Engineer",
                "name": "Duong Pham",
                "telephone": "0978100461",
              }
            }
            updateDeveloperInformation={[MockFunction]}
          />
        </GridItem>
        <GridItem>
          <EnhanceChangePasswordForm
            changePassword={[MockFunction]}
            email="test@gmail.com"
          />
        </GridItem>
      </Grid>
    </FlexContainerResponsive>
  </Content>
</FlexContainerBasic>
`;
